{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3387</th>\n",
       "      <td>MITE_T_9379|chr5A|253680838|253681189|GC|364|F534</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>90.56</td>\n",
       "      <td>339.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>213865.0</td>\n",
       "      <td>214203.0</td>\n",
       "      <td>2.000000e-129</td>\n",
       "      <td>464.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5079</th>\n",
       "      <td>MITE_T_11550|chr4B|491446578|491446917|ATTC|34...</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>94.86</td>\n",
       "      <td>331.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>337.0</td>\n",
       "      <td>213868.0</td>\n",
       "      <td>214198.0</td>\n",
       "      <td>3.000000e-146</td>\n",
       "      <td>520.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6222</th>\n",
       "      <td>MITE_T_13916|chr3D|469844247|469844577|GATA|34...</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>93.52</td>\n",
       "      <td>324.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>325.0</td>\n",
       "      <td>213868.0</td>\n",
       "      <td>214191.0</td>\n",
       "      <td>7.000000e-136</td>\n",
       "      <td>486.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6028</th>\n",
       "      <td>MITE_T_13765|chr7B|255748631|255748962|TTTT|33...</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>91.08</td>\n",
       "      <td>325.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>328.0</td>\n",
       "      <td>214192.0</td>\n",
       "      <td>213868.0</td>\n",
       "      <td>1.000000e-126</td>\n",
       "      <td>455.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6901</th>\n",
       "      <td>MITE_T_16464|chr6B|25945785|25946111|AA|329|F696</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>93.56</td>\n",
       "      <td>326.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>214193.0</td>\n",
       "      <td>213868.0</td>\n",
       "      <td>5.000000e-138</td>\n",
       "      <td>493.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 qseqid sseqid  pident  \\\n",
       "3387  MITE_T_9379|chr5A|253680838|253681189|GC|364|F534  chr1A   90.56   \n",
       "5079  MITE_T_11550|chr4B|491446578|491446917|ATTC|34...  chr1A   94.86   \n",
       "6222  MITE_T_13916|chr3D|469844247|469844577|GATA|34...  chr1A   93.52   \n",
       "6028  MITE_T_13765|chr7B|255748631|255748962|TTTT|33...  chr1A   91.08   \n",
       "6901   MITE_T_16464|chr6B|25945785|25946111|AA|329|F696  chr1A   93.56   \n",
       "\n",
       "      length  mismatch  gapopen  qstart   qend    sstart      send  \\\n",
       "3387   339.0      30.0      1.0    12.0  348.0  213865.0  214203.0   \n",
       "5079   331.0      17.0      0.0     7.0  337.0  213868.0  214198.0   \n",
       "6222   324.0      20.0      1.0     3.0  325.0  213868.0  214191.0   \n",
       "6028   325.0      29.0      0.0     4.0  328.0  214192.0  213868.0   \n",
       "6901   326.0      21.0      0.0     1.0  326.0  214193.0  213868.0   \n",
       "\n",
       "             evalue  bitscore  \n",
       "3387  2.000000e-129     464.0  \n",
       "5079  3.000000e-146     520.0  \n",
       "6222  7.000000e-136     486.0  \n",
       "6028  1.000000e-126     455.0  \n",
       "6901  5.000000e-138     493.0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#order blast by sseqid start\n",
    "import pandas as pd\n",
    "blast_file = 'extras/data/iwgsc_cn4/blast_90_90.csv'\n",
    "blast_file_orderer = 'extras/data/iwgsc_cn4/blast_90_90.order.csv'\n",
    "df_blast = pd.read_csv(blast_file, index_col=False, sep='\\t')\n",
    "df_blast.columns = ['qseqid','sseqid','pident','length','mismatch','gapopen','qstart','qend','sstart','send','evalue','bitscore']\n",
    "df_blast = df_blast.sort_values(['sseqid', 'sstart'])\n",
    "df_blast.to_csv(blast_file_orderer, sep=\"\\t\", index=None)\n",
    "df_blast.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>820</th>\n",
       "      <td>MITE_T_11550|chr4B|491446578|491446917|ATTC|34...</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>95.17</td>\n",
       "      <td>331</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>337</td>\n",
       "      <td>442471</td>\n",
       "      <td>442801</td>\n",
       "      <td>8.000000e-148</td>\n",
       "      <td>526.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2934</th>\n",
       "      <td>MITE_T_45970|chr3B|651188527|651188767|AT|42|F...</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>97.08</td>\n",
       "      <td>240</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>240</td>\n",
       "      <td>674129</td>\n",
       "      <td>674367</td>\n",
       "      <td>2.000000e-109</td>\n",
       "      <td>398.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2963</th>\n",
       "      <td>MITE_T_45970|chr3B|651188527|651188767|AT|42|F...</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>96.67</td>\n",
       "      <td>240</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>240</td>\n",
       "      <td>676678</td>\n",
       "      <td>676916</td>\n",
       "      <td>2.000000e-108</td>\n",
       "      <td>394.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2988</th>\n",
       "      <td>MITE_T_45970|chr3B|651188527|651188767|AT|42|F...</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>96.25</td>\n",
       "      <td>240</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>240</td>\n",
       "      <td>679227</td>\n",
       "      <td>679465</td>\n",
       "      <td>1.000000e-106</td>\n",
       "      <td>389.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14809</th>\n",
       "      <td>MITE_T_100638|chr4A|274958954|274959060|ACT|25...</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>95.28</td>\n",
       "      <td>106</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>106</td>\n",
       "      <td>712631</td>\n",
       "      <td>712526</td>\n",
       "      <td>6.000000e-41</td>\n",
       "      <td>168.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  qseqid sseqid  pident  \\\n",
       "820    MITE_T_11550|chr4B|491446578|491446917|ATTC|34...  chr1A   95.17   \n",
       "2934   MITE_T_45970|chr3B|651188527|651188767|AT|42|F...  chr1A   97.08   \n",
       "2963   MITE_T_45970|chr3B|651188527|651188767|AT|42|F...  chr1A   96.67   \n",
       "2988   MITE_T_45970|chr3B|651188527|651188767|AT|42|F...  chr1A   96.25   \n",
       "14809  MITE_T_100638|chr4A|274958954|274959060|ACT|25...  chr1A   95.28   \n",
       "\n",
       "       length  mismatch  gapopen  qstart  qend  sstart    send         evalue  \\\n",
       "820       331        16        0       7   337  442471  442801  8.000000e-148   \n",
       "2934      240         6        1       1   240  674129  674367  2.000000e-109   \n",
       "2963      240         7        1       1   240  676678  676916  2.000000e-108   \n",
       "2988      240         8        1       1   240  679227  679465  1.000000e-106   \n",
       "14809     106         5        0       1   106  712631  712526   6.000000e-41   \n",
       "\n",
       "       bitscore  \n",
       "820       526.0  \n",
       "2934      398.0  \n",
       "2963      394.0  \n",
       "2988      389.0  \n",
       "14809     168.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#order blast by sseqid start\n",
    "import pandas as pd\n",
    "blast_file = 'extras/data/iwgsc_cn4/blast_95.csv'\n",
    "blast_file_orderer = 'extras/data/iwgsc_cn4/blast_95_95.order.csv'\n",
    "df_blast = pd.read_csv(blast_file, index_col=False, sep='\\t')\n",
    "df_blast.columns = ['qseqid','sseqid','pident','length','mismatch','gapopen','qstart','qend','sstart','send','evalue','bitscore']\n",
    "df_blast = df_blast.sort_values(['sseqid', 'sstart'])\n",
    "df_blast.to_csv(blast_file_orderer, sep=\"\\t\", index=None)\n",
    "df_blast.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Most common MITEs in wheat / TREP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distance from MITEs to genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_file = '../data/iwgsc_refseqv1.0_HighConf_2017Mar13.gff3'\n",
    "blat_file = 'extras/data/iwgsc_cn4/blat_95_95_new.csv'\n",
    "blast_file = 'extras/data/iwgsc_cn4/blast_95.csv'\n",
    "genome_file = 'extras/data/iwgsc.gff3'\n",
    "\n",
    "#for rice data\n",
    "#genome_file = 'extras/data/rice.gff3.csv'\n",
    "#blat_file = 'extras/data/tracker/blat_90_90.csv'\n",
    "#gene_file = '../data/IRGSP-1.0_representative/locus.gff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load MITEs\n",
    "import pandas as pd\n",
    "import sys\n",
    "df_blat = pd.read_csv(blat_file, index_col=False, sep='\\t')\n",
    "df_blat = df_blat[df_blat.tName != 'chrUn'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load MITEs in blast\n",
    "import pandas as pd\n",
    "import sys\n",
    "df_blat = pd.read_csv(blast_file, index_col=False, sep='\\t')\n",
    "df_blat.columns = ['qseqid','sseqid','pident','length','mismatch','gapopen','qstart','qend','sstart','send','evalue','bitscore']\n",
    "df_blat = df_blat[df_blat.sseqid != 'chrUn'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load genes\n",
    "df_genes = pd.read_csv(gene_file, index_col=False, sep='\\t')\n",
    "df_genes.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes = df_genes[df_genes.feature == 'gene'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load genome structure\n",
    "df_genome = pd.read_csv(genome_file, index_col=False, sep='\\t', header=None)\n",
    "df_genome.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes_genome = {}\n",
    "for k,chromosome in df_genome.iterrows():\n",
    "    df_genes_genome[chromosome.seqname] = df_genes[(df_genes.seqname == chromosome.seqname)].reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "682396\n"
     ]
    }
   ],
   "source": [
    "#calculate distance to closest gene\n",
    "distances = []\n",
    "print(len(df_blat.index))\n",
    "for k,v in df_blat.iterrows():\n",
    "    midd = (v.sstart + v.send) / 2\n",
    "    df_curr = df_genes_genome[v.sseqid]\n",
    "    df_res = df_curr.loc[(df_curr.start - midd).abs().argsort()[:1]]\n",
    "    if len(df_res.index) > 0:\n",
    "        closest = (int(df_res.start) + int(df_res.end)) / 2\n",
    "        distance = abs(midd - closest)\n",
    "        distances.append(distance)\n",
    "    else:\n",
    "        print(v.qseqid, midd)\n",
    "        print(df_res)\n",
    "        print(df_curr)\n",
    "        print('error')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "3631676.5\n"
     ]
    }
   ],
   "source": [
    "#calculate distance max and min\n",
    "print(min(distances))\n",
    "print(max(distances))\n",
    "distances_kb = []\n",
    "for d in distances:\n",
    "    distances_kb.append(d/1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#! pip3 install seaborn\n",
    "distances_kb_filtered = []\n",
    "for distance_kb in distances_kb:\n",
    "    if distance_kb <= 500:\n",
    "        distances_kb_filtered.append(distance_kb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZ8AAAEZCAYAAABICyhRAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xm4HFWd//H3JwQUAiSBgSCEsBkQNBjigKCOBJRFFkEUWVTC4jaAMOrPYXEJyLhEB0VkBGQYCChEBAXiAIkINwwOW4CwGCBB1iQkLEkIYZQt398f53RSabr7dpLb1Ze+n9fz9HOrTlWdOnVud3/7nDpVpYjAzMysTP3aXQAzM+t7HHzMzKx0Dj5mZlY6Bx8zMyudg4+ZmZXOwcfMzErn4NPHSNpM0hJJ/fL8dZI+1+5y9Ta5jrZsdzl6szLeO/4/dC4HnxaT9ISkv0taryp9Wv5gDcvzF0n6rqQPSXpJ0iJJi/M6iwppQyV1Sfpbnq+8rlmBYi29uCsi9omIS5s4jl73JSBpjKT/aVH2bbkATtKukp5ux75XVLPvnVXdTSsylTRW0iWtyNua4+DTegE8DhxWSZD0HuDt1PhgRcStEbFORKwLvDuvM7CSFhGzctqxeb7yOqCE4+htROvKpRbl28x+e2Ndt0u7/g/WYg4+5bgUGFOYHwOMX4Hta30Am/pQSuon6d8lPSfpUWDfquU3Szo6T2+VW1ULJT0r6fKcPiXv7/7cyjpY0iBJE/N6L+TpTary/a6kW/M2NxRbf7mF92dJCyQ9KemInL5GLu+Tkp6R9AtJb6txXO8CzgV2ya3C+Tl9XUmX5HI9Lumb3dTNqZIelfSipLuKx1BYr26e9eqsUkZJk3P9PCTp4MKyfST9JdfN05K+Jmkt4Dpg40JLd6Ma5Vkv1/eLku6QdEaxBdjNfi+SdI6kP+T8b5O0RTPb1ihH8b0zRtL/SPqxpPmS/ipp7zrbHSnp2sL8o5ImFOafkrR9YZM9JM3IZTqnKq+jJU3Py65X7knIy87KeVX+tx/K6XsBpwKH5Hq+t045R0m6J29/haQJkr5bWL6fpHvze/hWSSMKyx6X9HVJ9+Xll0tao7D8C5JmSnpe0tWS3lFYtkTSP+djflHpc7SlpP/N77MJkvo3U45eLSL8auGL1OrZHXgI2IYU8J8ENgWWAMPyehcB363adjPgDaBfVfrNwNFN7v/LwHRgY2AQcFMxz2JewGXAKXl6DeADhXyWAFsU5tcDPgG8DRgA/Ab4fVUZZwJb5XVuBr6flw0DFgGfBlYDBgPb52VnAVcDA3O+1wDfq3NsY4BbqtIuAX4PrJXr7xHgqDrbfwO4D3hnnh8BDC4c75bd5VmvzvK6TwFHkAL3SOA5YNu8fE5h3YHAyDy9K/BUN//TCXm/bwO2zfu5pcn9XgQ8D7yP9F78FXBZg22frWxboxzF984Y4BXg6Lztl4HZdbbbApifpzcCngCezvNbAi9Uve+uBdYhfWaeBfbMyw4EZgBb52M5FfhzYdvDSe/5fsBXgWeANfKyscAlDep49Vyu40nv0U/k4/tuXj4KmAf8Yz7ez5E+66sXPve3A0NyGaYDX8zLds//k/fm/ZwNTKk65qtJ7/9tgb8DfyS999YB/gJ8rply9OZX2wvQ6S+WBZ9Tge8DewGT8ht6VYLPy8B8YEH+e3qd/f+p8qbP83tQP/iMB84DNqmRz9Iv4zr7GVn1pXEzcGph/p+B6/L0ycBVdfJZzPJBbhfgsTrrLhd88pfM34FtCmlfBG6qs/3DwH51li0hfRE2zLNenZEC65SqtPOAb+fpJ4AvAOtUrdMw+OTyvEoOmDntDJYFn+72exHwy8KyjwHTm9m2Rlmqg8+MwrI18/tswzrbPpnfM4cA55O+qLcGjgSurvo/7FKY/w3wr3n6Ogo/LHLdvAxsWmef84ERebq74PNP5IBYSPsflgWfX1D1mcvvp3/K048DhxWWjQN+kaf/E/hhYdmA/D8dVjjmnQvLpwLfKMz/O/CTZsrRm1/udivPr0i/xI4k/ZJeVV+JiPUiYnD+O7bOehsDxRPYTzbI8xukD/Cdkh6QdFS9FSWtKel8pQEVC4EpwCBJxe7AuYXp/wPWztObAn+tkecGpF/fd+eum/nA9cD6Dcpc9A+kX5JPFdKeBN7UlVYox2OrmOe/UrvONgN2rhyHpAWk//+QvPyTpC7QJ3P31c7dHl2yAemHy6xCWvH/291+of7/pd62b+r6q2NpvhHxN9Iv8bXrrDsF2A34MNCVX6NJwXdK1brzGpT3Z4X3yguk82WbAORur+m5O2oBsC7p/9mMjYHZVWnV9fz1qroamrfrrtwbU/gcRsTLuezF9+mzhem/VeX1N5avg+7K0Sv1734V6wkR8ZSkx0m/NI8ucdfPkL5kKzart2JEPEv6VY+kDwI3SpoSEbW+oL8ODAd2jIjnJL0XuIfmTpg/DexUI/150of03RHxTDd5UGM/zwOvkY7x4Zy2GW/+Eql4itQtOL3BPhrmGRHzqFFnpGPsioi9ahY84m7gQEmrAV8BriB1R3ZXd88Br5O+YB7NacX/b8P9dmNVtl1RtwD7A5sD3wNeBD4D7Az8vMk8ngb+LSIur16Qz+/8K7BbREzPafNZdq60u3p+hjf/aNmUZXX+NKk7+AdNlrVoDoXPoaQBpB9Ys+puUd+qlKOt3PIp19HA7vlXYbNWdbTPFcAJkjaRNBg4qe6OpE9p2Qn3haTm/xt5fi6pG6piHdIvsEVKAwlOW4Ey/Rr4SN7fakon0N8bqc/gAuCs3Aoil3vPOvnMA4ZKWh0gIpbk4/2epLUlbUbq6683HPhC4AxJ78z7GpHraKnu8mxQZ38Atpb0WUn9Ja0u6R+VTuivLulwSetGxBvAS6SAUjmm9SWtW6vAuTy/A07Lrc93kc7RVNTb7zZ16qCobpmb2HZFVVo+a0bEHFKX1t6kL+GaAwBqOA84VdJ2AJIGSvpUXrYO6UfDC0qDWL6T0yrmAZtXtdSLbgPekHRcfo8ewPI/mC4Avixpp7zvAUqDSAY0Ue7LgKMkba80mOb7wO0RsTJD7FelHG3l4NN6S39hRcTjEXFPrWXNbF/lHC27xuclSXfVWe8C0jmm+0h9x1c1yH9H4A5Ji0gnPE+IiEr3wGnAJblp/yngp6QusueB/yX1vzdTbvKHbB/g/5H64e8FKqObTib9urw9d+dNJp0LqOUm0snXuZIq3RQnkFpPj5F+Xf8qIi6qs/1PSIFlsqQXSX3xa9Yof6M8a9ZZRCwG9gQOJf3SnQP8kDQoAfKJ4XyMXwQ+m+vmEeBy4LFc17W6vL5COon9DOmc02Wkk+E02O+bRgxWa6LMb9qkuywb7GsmKejekudfInXF3pp/hNTLo/h5ujqXb0Kux/tJAQzSe/4G0oCEx0n/v+KX+29JP+xekDS1RvleAw4CPk86r3o4MJFl9Xw36ZzdOblFNYPlR7Q2OvabgG+TfkTMJg3AOLSZY66RV3fl6LW0/P+5hzOXhpLOb2xE+jX4y4j4uaSxpAqrfGGcGhE35G1OIbUQXgdOjIjJOX1v0kiofsCFETEup29OGv0zmNTt87mIeF1pWOMlpFE9zwOHRESx396sI0j6ITAkIuqeo7NVJ+l24NyIWJHLJKyOVrd8Xge+FhHbkUYtHV9owv8kIkblVyXwbEsacbMt6dzIL5T0A84hjRR7N3BYIZ9xwJkRsQ2p2+OYnH4MaTjncFLQ+lGLj9WsFJK2Ub6WI3e3HEP6FW09SNKHJQ3J3W5jSEPxb2h3uTpFS4NPRMyNiGl5ejHpWpdK/3itvtYDgAkR8XpEPEG6TmSn/JqZuzNeI7V0Klf0786yrqTxpLH/lbwqv1CuBD7SU8dl1mbrAL+TtJj0WfhxRExsc5k60Tak7uqFpPN8n8wDTKwHlDbaLXePjQTuAD4EHKd0U8KpwNcj4kVSYLqtsNnsnCaW76+dBewkaX1gQT4JW0mvBLdNKttExBtKVwavFxHzW3B4ZqWJiKmkkYbWQhFxAemcqbVAKQMOJK1Nan2cmFtAvwC2ioiRpFFUZ1ZWrbF5dJNevaxyEqs63ffMMjPrJVre8lG6B9GVwKURcQ1ARDxXWOUC0igSSC2X4jULQ0kjbkS6BmK59Ih4XukeY/1y66eyfjGvOflainUjYkGN8jkgmZmthIhY6UtBymj5/Bfp9h0/qyRUDR89CHgwT18LHJrH5W8BvBO4E7gLeKfSs2jWIA1LrDxC4CagcvPDMYX0a1k25PDgvF5N7b7NRG95jR07tu1l6C0v14XrwnXR+LWqWtryyVd8fwZ4QOnOsUG6x9nhkkaSLsh7AvgSQERMl3QF6Yrz10iPDQjSxV7Hk675qAy1rlxtfjJpnP8ZpOtFLszpFwKXSppJunVFcRy9mZm1UUuDT0T8mXQfqmp1hytGuk3Em24VEWk49puu0o6Ix4H310h/hTRs28zMehnf4cCWGj16dLuL0Gu4LpZxXSzjuug5Lb3DwVuBpOjrdWBmtqIkEb18wIGZmdlyHHzMzKx0Dj5mZlY6Bx8zMyudg4+ZmZXOwcfMzErn4GNmZqVz8DEzs9I5+JiZWekcfMzMrHQOPmZmVjoHHzMzK52Dj5mZlc7Bx8zMSufgY2ZmpXPwMTOz0jn4mJlZ6Rx8zMysdA4+ZmZWOgcfMzMrnYOPmZmVzsHHzMxK5+BjZmalc/AxM7PSOfiYmVnpHHzMzKx0Dj5mZlY6Bx8zMyudg4+ZmZXOwcfMzErXv90F6A0efPDBpdNDhw5l0KBBbSyNmVnnc/ABJkxYBMDLLy9gjz2eY599dmtziczMOpuDDzBs2AcAmDdvBkuWzG5zaczMOp/P+ZiZWelaGnwkDZV0k6Tpkh6QdEJOHyxpsqRHJE2SNLCwzdmSZkqaJmlkIX2MpBl5myMK6aMk3Z+XnVVIr7sPMzNrr1a3fF4HvhYR2wG7AMdJehdwMnBjRGwD3AScAiDpY8BWETEc+BJwXk4fDHwH2BF4PzC2EEzOBT4fEVsDW0vaK6fX3IeZmbVfS4NPRMyNiGl5ejHwEDAUOAAYn1cbn+fJfy/J698BDJQ0BNgLmBwRL0bEQmAysLekjYB1IuLOvP0lwIGFvIr7qKSbmVmblXbOR9LmwEjgdmBIRMyDFKCADfNqmwBPFzabldOq02cX0mfVWJ8a+9ig547GzMxWRSmj3SStDVwJnBgRiyVFvVVrzEeNdLpJXyETJ54GwOLFL9C//5bst5+HWpuZFXV1ddHV1dVj+bU8+EjqTwo8l0bENTl5nqQhETEvd509m9NnAZsWNh8KzMnpo6vSb26wPsDcOvt4k/33Py0Vat4MRozwUGszs2qjR49m9OjRS+dPP/30VcqvjG63/wKmR8TPCmnXAkfm6SOBawrpRwBI2hlYmLvOJgF7SBqYBx/sAUzK3WmLJO0kSXnba2rsY0wh3czM2qylLR9JHwQ+Azwg6V5Sl9ipwDjgCklHA08BBwNExHWS9pH0KPAycFROXyDpDGBqzuP0PPAA4FjgYuDtwHURcUNOr7kPMzNrv5YGn4j4M7BancUfrbPN8XXSLyYFmer0u4ERNdLn19uHmZm1l+9wYGZmpXPwMTOz0jn4mJlZ6Rx8zMysdA4+ZmZWOgcfMzMrnYOPmZmVzsHHzMxK5+BjZmalc/AxM7PSOfiYmVnpHHzMzKx0Dj5mZlY6Bx8zMyudg4+ZmZXOwcfMzErn4GNmZqVz8DEzs9I5+JiZWekcfMzMrHQOPmZmVrpug4+kAZL65emtJX1c0uqtL5qZmXWqZlo+twBvl7QJ8CfgKODiVhbKzMw6WzPBRxHxf8BBwM8j4hPAdq0tlpmZdbKmgo+kXYDPAP+d0/q3rkhmZtbpmgk+JwKnAL+PiL9I2hK4ubXFMjOzTtZtCyYibiGd96nMPybpay0tlZmZdbS6LR9JtxamL61afGfLSmRmZh2vUbfbgML0e6qWqQVlMTOzPqJR8Ik607XmzczMmtbonM8gSZ8gBahBkg7K6QIGtrxkZmbWsRoFnynAxwvT+xeW3fLm1c3MzJrTKPhMjIjflVYSMzPrMxqd8/lWaaUwM7M+xXe1NjOz0jUKPu+SdH+N1wOS7m8mc0kXSppXXF/SWEmzJN2TX3sXlp0iaaakhyTtWUjfW9LDkmZIOqmQvrmk2yU9IulySf1z+hqSJuS8bpM0bIVqxczMWqrROZ/HWX6Qwcq4CPg5cElV+k8i4ifFBEnbAp8GtgWGAjdKGk4aXXcO8BFgDnCXpGsi4mFgHHBmRPxW0rnAMcD5+e/8iBgu6RDgR8Chq3gsZmbWQxq1fF6NiCfrvZrJPCJuBRbUWFTrItUDgAkR8XpEPAHMBHbKr5l5v68BE/K6ALsDV+Xp8cCBhbzG5+krSYHLzMx6iUbB588t3O9xkqZJ+k9JlWuGNgGeLqwzO6dVp88CNpG0PrAgIpYU06vziog3gIWS1mvNoZiZ2Ypq2O3W6Aai1d1mK+AXwHcjIiT9G3Am8Hlqt4aC2gEy8vrV21TuvFCdLnxXBjOzXqNR8Fm7FTuMiOcKsxcAE/P0LGDTwrKhpHM8AoZVp0fE85IGSeqXWz+V9Yt5zZG0GrBuRNTq/gNg4sTTAFi8+AX699+S/fbbbWUPz8ysI3V1ddHV1dVj+SmitQ0CSZuTLlgdkec3ioi5efqrwI4Rcbik7YBfA+8ndZv9ERhOavk8Qjpv8wzpjtqHRsTDkn4D/C4ifpMHHNwXEedJOhZ4T0QcK+lQ4MCIqDngQFKcf36qg3nzZrDDDrMdfMzMuiGJiFjpm0zXbflIOrvRhhFxQneZS7oMGA2sL+kpYCywm6SRwBLgCeBLOb/pkq4ApgOvAcdGioxvSDoemEwKRBfmkW4AJwMTJJ0B3AtcmNMvBC6VNBN4AY90MzPrVeq2fCS9CjwIXMGy7q+lImJ8re3eatzyMTNbcS1r+QDvAA4GDgFeB34DXNXo3ImZmVkz6g61jogXIuK8iNgNOBIYBPxF0ufKKpyZmXWmRi0fACSNAg4D9gCuB+5udaHMzKyzNRpwcDqwH/AQ6a4Cp0TE62UVzMzMOlejls+3gceA9+bX9yVBvmAzIrZvffHMzKwTNQo+W5RWCjMz61PqBp9mbx5qZma2ovwwOTMzK52Dj5mZla5u8JH0p/x3XHnFMTOzvqDhHQ4k7Qp8XNIE3nx7nXtaWjIzM+tYjYLPd0g37hwKVD+7J0hPETUzM1thjUa7XQlcKenbEXFGiWUyM7MO1+3tdSLiDEkfBz6ck7oi4g+tLZaZmXWybke7SfoBcCLpOTvTgRNzmpmZ2UrptuUD7AuMzI+qRtJ40oPbTmllwczMrHM1e53PoML0wFYUxMzM+o5mWj4/AO6VdDNpuPWHcavHzMxWQTMDDi6X1AXsSAo+J0XE3FYXzMzMOlczLR8i4hng2haXxczM+gjf283MzErn4GNmZqVrGHwk9ZP0YFmFMTOzvqFh8MnX9twnaVhJ5TEzsz6gmQEH7wD+IulO4OVKYkR8vGWlMjOzjtZM8Dm95aUwM7M+pZnrfKZI2gwYHhE3SloLWK31RTMzs07VzI1FvwBcCZyfkzYBrm5loczMrLM1M9T6OOCDwCKAiJgJbNjKQpmZWWdrJvi8EhGvVmYk9Sc9ydTMzGylNBN8pkg6FVhT0h7Ab4GJrS2WmZl1smaCz8nAc8ADwJeA64BvtbJQZmbW2ZoZ7bYkP0DuDlJ32yMR4W43MzNbad0GH0n7AucBfyU9UmELSV+KiOtbXTgzM+tMzVxkeiawW0Q8CiBpK+C/AQcfMzNbKc2c83m2Eniyx4BnW1QeMzPrA+oGH0kHSTqIdF+36yQdKWkMaaTbXc1kLulCSfMk3V9IGyxpsqRHJE2SNLCw7GxJMyVNkzSykD5G0oy8zRGF9FGS7s/LzmpmH2Zm1n6NWj7759fbgXnArsBo0si3wU3mfxGwV1XaycCNEbENcBNwCoCkjwFbRcRw0qi683L6YOA7pMd4vx8YWwgm5wKfj4itga0l7dVoH2Zm1jvUPecTEUetauYRcWu+L1zRAaRABjAeuJkULA4ALsnb3SFpoKQhwG7A5Ih4EUDSZGBvSVOAdSLizpzXJcCBwKQa++jK+zAzs16gmdFuWwBfATYvrr8Kj1TYMCLm5TzmSqrcqmcT4OnCerNyWnX67EL6rBrrAwyp2scGK1lWMzNrgWZGu10NXEg617OkhWVRjfmokU436WZm1ss1E3z+HhFn9+A+50kaEhHzJG3EspFzs4BNC+sNBebk9NFV6Tc3WB9gbp191DRx4mkALF78Av37b8l+++22MsdlZtaxurq66Orq6rH81N3NCiQdDgwHJgOvVNIj4p6mdiBtDkyMiBF5fhwwPyLGSToZGBQRJ0vaBzguIvaVtDNwVkTsnAccTAVGkQZITAXeFxELJd1B6hK8i3Tt0dkRcUPVPk4CBkdEzXM+kuL881MdzJs3gx12mO3gY2bWDUlERK0eqKY00/IZAXwO2J1l3W6R57sr3GWkVsv6kp4CxgI/BH4r6WjgKeBggIi4TtI+kh4lPa77qJy+QNIZpKATwOkRsTDv4ljgYtKIvOsi4oacPg64onofZmbWOzTT8nkY2L74WIVO4paPmdmKW9WWTzN3OLgPGLSyOzAzM6vWTLfbEOBhSXex/DmflR1qbWZmfVwzwWdsy0thZmZ9SjPP85lSRkHMzKzvaOYOBy+x7OLNNYDVgZcjYt1WFszMzDpXMy2fdYrzkg4EdmpZiczMrOM1M9ptORFxNU1c42NmZlZPM91uBxVm+wH/iO+hZmZmq6CZ0W77F6ZfB54gPbLAzMxspTRzzmeVn+tjZmZWVDf4SPpOg+0iIs5oQXnMzKwPaNTyeblG2gDgGGB9wMHHzMxWSqPHaJ9ZmZa0DnAi6U7TE4Az621nZmbWnYbnfCStB3wN+AwwHhgVEQvKKJiZmXWuRud8fgwcBPwSGBERi0srlZmZdbRGF5l+HdgY+BYwR9Ki/HpJ0qJyimdmZp2o0TmfFb77gZmZWTMcYMzMrHQOPmZmVjoHHzMzK52Dj5mZlc7Bx8zMSufgY2ZmpXPwMTOz0jn4mJlZ6Rx8zMysdA4+ZmZWOgcfMzMrnYOPmZmVzsHHzMxK5+BjZmalc/AxM7PSOfiYmVnpHHzMzKx0Dj5mZlY6Bx8zMytd24KPpCck3SfpXkl35rTBkiZLekTSJEkDC+ufLWmmpGmSRhbSx0iakbc5opA+StL9edlZ5R6dmZk10s6WzxJgdETsEBE75bSTgRsjYhvgJuAUAEkfA7aKiOHAl4Dzcvpg4DvAjsD7gbGFgHUu8PmI2BrYWtJeJR2XmZl1o53BRzX2fwAwPk+Pz/OV9EsAIuIOYKCkIcBewOSIeDEiFgKTgb0lbQSsExF35u0vAQ5s2ZGYmdkKaWfwCWCSpLskfT6nDYmIeQARMRfYMKdvAjxd2HZWTqtOn11In1VjfTMz6wX6t3HfH4iIuZI2ACZLeoQUkGpRjfmokU436TVNnHgaAIsXv0D//luy3367dVN0M7O+pauri66urh7Lr23BJ7dsiIjnJF0N7ATMkzQkIublrrNn8+qzgE0Lmw8F5uT00VXpNzdYv6b99z8NgHnzZjBixOyVPygzsw41evRoRo8evXT+9NNPX6X82tLtJmktSWvn6QHAnsADwLXAkXm1I4Fr8vS1wBF5/Z2Bhbl7bhKwh6SBefDBHsCkHNgWSdpJkvK2lbzMzKzN2tXyGQL8XlLkMvw6IiZLmgpcIelo4CngYICIuE7SPpIeBV4GjsrpCySdAUwldaudngceABwLXAy8HbguIm5opmB/+EMXt902c+n8xhuvzXHHHb7KB2xmZsu0JfhExOPAyBrp84GP1tnm+DrpF5OCTHX63cCIFS3b/PmvMWrUF5fOP/nkL1c0CzMz64bvcGBmZqVz8DEzs9I5+JiZWekcfMzMrHQOPmZmVjoHHzMzK52Dj5mZlc7Bx8zMSufgY2ZmpXPwMTOz0jn4mJlZ6Rx8zMysdA4+ZmZWOgcfMzMrnYOPmZmVzsHHzMxK164nmb5lTJ06jW9+c/kHyvnppmZmq8bBpxuLFy9hs82+uFyan25qZrZq3O1mZmalc/AxM7PSudttJVSfB/I5IDOzFePgsxKqzwNdddWxzJmzeOm8g5GZWWMOPj2gu2A0c+ZDDB++7dJ5Bycz6+scfFqgOhh1dX2Zj37ULSUzswoHnzaoDk61hm7/x39c5gBlZh3LwacXqHUh69SpD/LJT569dN7XFplZJ3Hw6QVqXcja1fXl5eY9ws7MOomDz1uER9iZWSdx8HmL6u68UfU5I3CAMrPew8GnQ1R3y1WfMwKfNzKz3sPBp0PUGt5drTpA+fojM2sXB58+ZEWvP3JwMrNWcfCxpVY1OIEDlJk1x8HHmtZdcAK3nsysOQ4+1qNWtfXk1pRZ39DRwUfS3sBZpOcWXRgR49pcpD6vu+C0Mq2p7uYdvMx6n44NPpL6AecAHwHmAHdJuiYiHm5vyXqvF1+c0+4i1LSiAWtlzlVVpy1e/BQ/+9m/LZ3vy/fa6+rqYvTo0e0uRq/guug5HRt8gJ2AmRHxJICkCcABgINPHb01+KyqZs5VVaeNG7cja689bOl89XVTKxPQumuR9dYA5y/cZVwXPaeTg88mwNOF+VmkgGTWrVdfjYbXTa1MQOuuRdaKALcyXZLVQfDqq2/gtdc2XqU8zap1cvBRjbSoteLTT98AwN///jL9+tXazKzndXdhcE8EuBUNgPDmILho0S9XqNuzVp6rGiSbGZhSRp4rGoh7S7lbsY9VpYia38dveZJ2Bk6LiL3z/MlAVA86kNSZFWBm1mIRsdK/1js5+KwGPEIacPAMcCdwWEQ81NaCmZlZ53a7RcQbko4HJrNsqLUDj5lZL9CxLR8zM+u9+rW7AO0iaW9JD0uaIemkdpenDJIulDRP0v2FtMGSJkt6RNIkSQMLy86WNFPSNEkj21PqnidpqKSbJE2X9ICkE3J6X6yLt0m6Q9K9uS7G5vTNJd2e6+JySf1z+hqSJuS6uE3SsMZ7eOuR1E/SPZJMBpTTAAAJWklEQVSuzfN9si4kPSHpvvzeuDOn9dhnpE8Gn8IFqHsB7wYOk/Su9paqFBeRjrnoZODGiNgGuAk4BUDSx4CtImI48CXgvDIL2mKvA1+LiO2AXYDj8v+/z9VFRLwC7BYROwAjgY9Jej8wDjgz18VC4Ji8yTHA/FwXZwE/akOxW+1EYHphvq/WxRJgdETsEBGVy1R67jMSEX3uBewMXF+YPxk4qd3lKunYNwPuL8w/DAzJ0xsBD+Xp84BDCus9VFmv017A1cBH+3pdAGsBU0nXwz0L9MvpSz8vwA3A+/P0asBz7S53D9fBUOCPwGjg2pz2XB+ti8eB9avSeuwz0idbPtS+AHWTNpWl3TaMiHkAETEX2DCnV9fRbDqwjiRtTvrFfzvpw9Ln6iJ3M90LzCV98f4VWBgRS/Iqxc/H0rqIiDeAhZLWK7nIrfRT4BvkawIlrQ8s6KN1EcAkSXdJ+nxO67HPSMeOdutG0xeg9mEdX0eS1gauBE6MiMUNrvnq6LrIX6w7SFoX+D1Q62rCyvFW14XokLqQtC8wLyKmSRpdSebNx9zxdZF9ICLmStoAmCzpEeof3wp/Rvpqy2cWUDw5OJR089G+aJ6kIQCSNiJ1t0Cqo00L63VUHeWTxlcCl0bENTm5T9ZFRUQsAqaQupYG5XOjsPzxLq2LfC3duhGxoOyytsgHgY9Legy4HNiddC5nYB+si0rLhoh4jtQ1vRM9+Bnpq8HnLuCdkjaTtAZwKHBtm8tUlupfctcCR+bpI4FrCulHwNK7RSysNLc7xH8B0yPiZ4W0PlcXkv6hMmJJ0pqkc1/TgZuBg/NqY1i+Lsbk6YNJJ507QkScGhHDImJL0nfCTRHxWfpgXUhaK/cMIGkAsCfwAD35GWn3Sa02nkzbm3QHhJnAye0uT0nHfBnp18grwFPAUcBg4MZcF38EBhXWPwd4FLgPGNXu8vdgPXwQeAOYBtwL3JPfD+v1wboYkY9/GnA/8M2cvgVwBzAD+A2wek5/G3BF/tzcDmze7mNoUb3syrIBB32uLvIxVz4fD1S+I3vyM+KLTM3MrHR9tdvNzMzayMHHzMxK5+BjZmalc/AxM7PSOfiYmVnpHHzMzKx0Dj7WNpLeyLeufzDftv2rkpSXvU/SWQ223UzSYeWVdrl9D5T0zz2U11hJX+uJvBrso8fK21Mk/UnS2vn/+ECN5WMk/bxG+uqSphTuOGBvUf4HWju9HBGjIuI9wB7APsBYgIi4OyL+pcG2WwCHl1DGWgYDx7Zp3yujV5VX0j7AtIhYnJPqXWz4pvSIeI10keOhLSqelcTBx3qFiHge+CJwPICkXSVNLEzfm1tJd+fbffwA+FBOOzH/gr5F0tT82rmw7c2SfivpIUmXVvYpaUdJf84Pv7pd0oB8h+cfKT1gbZqkL9Qo7g+ALfO+x+W8fqz0MLb7JH261jFKOqLwcK7xNZaPzA8lmybpqsJtb06Q9JecfllOW0vp4YB35DrZP6dvl9PuyetvVau8Vfv9dq6bSZIuq7TEJG0p6fp8V+MpkrbO6RdJ+lmuu0clHVTI6/9JujPve2ydf/dnWHZblmI5KmV8X04alvf/kKTvFFa9Judhb2Xtvo2DX333BSyqkfYCsAHL397kWmCXPL0W6UfT0uU5/e3AGnn6ncBdeXpXYAHwDtI97f4X+ACwOunRAaPyemuTnsnyBeDUnLYG6T6Am1WVsfqZSAcBk/L0hsCTVD3LBNiO9IyTwXl+UP47lvRgO0i3JflQnj4d+Emens2yW7qsm/9+Dzg8Tw8k3e5kTeBs4LCc3p90C5jlyltVrveRbq+zRq6DGYXy3Eh6QBikm0r+KU9fBPwmT28LzMzTewDn52kBEyvHU7XPJ4ABxboEts7lGJHTx+TjHpT/tw8U/lf9gGfb/f71a9VeffWRCtZ71WqN/xn4qaRfA7+LiNn51FDRGsA5So/vfQMYXlh2Z0Q8AyBpGrA5sAiYExH3AETuApK0JzBCUuVGkuvmvJ5sUOYPke6CTEQ8K6kL2BH4Q2Gd3YErI9/1OCIWFjNQepzBwIi4NSeNJ903DFJQukzS1aS7C0O60eP+kr5ROP5hwG3ANyVtmuvq0Rp1VV32ayLiVeDVQmtzAClI/1bLMli9sN3V+TgeklR5psuewB6S7iEFnwGkuruV5Q2OiJcL8xvm/D4ZEQ8V0v9YqSdJv8tlvScilkh6RdKAqnzsLcTBx3oNSVsCr0fEc8UvzIgYJ+kPwL7A7ZI+UmPzrwJzI2J7pdvb/62w7JXC9Buk9329b2QBX4mIP65I0buZr6R1dyPFemXaF/gwcADwbUnvzut+MiJmVq37iKTbgf1IDwI7hvREyhXdZz/SQ9RG1VlerFMV/v4gIi5osD9IjzEvepH0ILIPkVqHFdX1VZx/G/D3bvZjvZjP+Vg7Lf3iU3pg1blArRFOW0bEXyLiR6THPL8LeInUKqkYCDyTp48gdaE18jDwjsr5hTzyajVgEnCs0vN+kDRc6VEDRS8B6xTmbwEOyeeLNgD+Cbizaps/AZ9WftKlpMHFhZGepTNf0gdz0udIz9YBGBYRU4CT8nEOyOU8obJ9bvEhaYuIeDwifk7qrty+RnmLbiW1oN6mdAv9fXN5XgIel/Spwj62r5NH5f84CTg6t5qQtHGuj2qP5B8aFa8ABwJHaPkRjHtIGpTr/0BSC5hch89FenqovUW55WPt9PbcRbMG8BpwSUT8tMZ6/yJpN9Iv5unA9aRfwa8rPf75YuA/gN/l7rIuoF53TEAaNSXpEFJX3ZrA/5GeZfOfpG65e3J307OkL75lGUTMzyfb7weuj4iTJO1C6h5bAnwjIp6t2ma6pO8BUyS9TrpV/dFVZTsSOC+X5zHgqBwEf5W75UQ6D7RI0hnAWbkMIrVuPk4Kgp/N9fkMcHpELKwub6FcUyVdS7p9/pOkc1wv5sWfBc6V9C3Sd8UE0vmZmi2SiPijpHcBt+WW60s5j+eq1v9vYLd8jJVy/E3SfqQnZlZGwd0K/ArYCvh1pYs0b3sd9pbmRyqY9XGVcyc56N0CfCEiprVwfxsB4yNir5Xc/irS82WquxztLcQtHzP7paTtSOdRLm5l4IH0eGZJF0haO5Zd69MUSasDv3fgeetzy8fMzErnAQdmZlY6Bx8zMyudg4+ZmZXOwcfMzErn4GNmZqVz8DEzs9L9f3YXSFT8BLDpAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1384f0358>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "# PLOT\n",
    "# Import the libraries\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# seaborn histogram\n",
    "sns.distplot(distances_kb_filtered, hist=True, kde=False, \n",
    "             bins=80, color = 'blue',\n",
    "             hist_kws={'edgecolor':'black'},rug=False)\n",
    "# Add labels\n",
    "plt.title('MITE distance to closest gene in wheat genome')\n",
    "plt.xlabel('Distance to closest gene (kb)')\n",
    "plt.ylabel('Number of MITEs')\n",
    "plt.savefig('extras/histogram.png', dpi=800,bbox_inches='tight')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Count Families"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count families \n",
    "import pandas as pd\n",
    "import sys\n",
    "df_blat = pd.read_csv('extras/data/iwgsc_cn4/blat_ts17_95_95.csv', index_col=False, sep='\\t')\n",
    "df_blat.columns = ['match','mismatch','repmatch','ns','qGapCount','qGapBases','tGapCount','tGapBases','strand','qName','qSize','qStart','qEnd','tName','tSize','tStart','tEnd','blockCount','blockSizes','qStarts','tStarts','cov','sa']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = df_blat.qseqid.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8027 MITE_T_90576|chr3D|581983465|581983588|TA|35|F3909\n",
      "4343 MITE_T_45970|chr3B|651188527|651188767|AT|42|F1918\n",
      "3927 MITE_T_119843|chr2B|409855987|409856068|TA|85|F5463\n",
      "3897 MITE_T_84430|chr3D|531147691|531147821|TA|43|F3641\n",
      "3879 MITE_T_76234|chr1D|360358225|360358385|TA|169|F3188\n",
      "3678 MITE_T_67707|chr4B|488893119|488893288|TT|176|F2742\n",
      "3627 MITE_T_113077|chr4B|190259909|190259996|TA|91|F5145\n",
      "3547 MITE_T_118365|chr5A|113012180|113012261|TA|85|F5422\n",
      "3546 MITE_T_112462|chr2D|335449673|335449760|TA|91|F5136\n",
      "3505 MITE_T_114717|chr1D|98479831|98479916|AG|85|F5233\n",
      "3494 MITE_T_116731|chr3D|27621554|27621637|AT|87|F5322\n",
      "3418 MITE_T_119032|chr5B|127018445|127018526|TA|85|F5434\n",
      "3048 MITE_T_74129|chr6B|320412940|320413102|TA|173|F3036\n",
      "3035 MITE_T_74920|chr3A|585605639|585605800|TA|173|F3090\n",
      "3001 MITE_T_88612|chr4B|499756008|499756133|AGT|132|F3813\n",
      "2996 MITE_T_74720|chr5A|206730853|206731014|TA|173|F3077\n",
      "2869 MITE_T_113343|chr2B|179285004|179285091|GA|87|F5153\n",
      "2855 MITE_T_114079|chr2A|11304740|11304827|AG|87|F5192\n",
      "2852 MITE_T_115150|chr5B|212481103|212481188|AG|85|F5237\n",
      "2834 MITE_T_67285|chr3D|360695824|360695994|AT|179|F2719\n",
      "2831 MITE_T_75868|chr6A|130354837|130354997|TA|172|F3158\n",
      "2795 MITE_T_74771|chr3B|621150041|621150202|TA|173|F3081\n",
      "2774 MITE_T_75328|chr7D|381797793|381797954|TA|173|F3112\n",
      "2768 MITE_T_68769|chr1D|345368677|345368845|TA|180|F2787\n",
      "2710 MITE_T_69392|chr5B|479312783|479312950|CTT|175|F2829\n",
      "2709 MITE_T_86948|chr4D|439129800|439129927|CA|129|F3744\n",
      "2674 MITE_T_75536|chr3D|185002608|185002769|TA|172|F3131\n",
      "2657 MITE_T_75960|chr2D|616648357|616648517|TA|171|F3167\n",
      "2642 MITE_T_66504|chr4B|479123258|479123428|TATA|187|F2708\n",
      "2621 MITE_T_71978|chr3D|340455280|340455443|AT|175|F2975\n",
      "2609 MITE_T_72475|chr3D|503092467|503092630|AT|174|F2988\n",
      "2598 MITE_T_88409|chr6D|335001133|335001258|AGT|132|F3806\n",
      "2556 MITE_T_86906|chr3D|261156870|261156997|TG|129|F3741\n",
      "2551 MITE_T_86394|chr3D|454698735|454698863|TATA|139|F3717\n",
      "2548 MITE_T_83087|chr3D|7232862|7232993|TA|139|F3602\n",
      "2537 MITE_T_73105|chr1D|256009888|256010051|AT|55|F2989\n",
      "2514 MITE_T_72317|chr7A|223092167|223092330|AT|55|F2982\n",
      "2486 MITE_T_82458|chr3B|679578880|679579013|TA|139|F3572\n",
      "2449 MITE_T_97364|chr7B|18251405|18251516|TA|41|F4299\n",
      "2326 MITE_T_98563|chr3D|588579729|588579840|TA|29|F4301\n",
      "2319 MITE_T_73309|chr5D|337983951|337984114|AT|174|F2994\n",
      "2269 MITE_T_101224|chr4B|111927577|111927681|TA|37|F4460\n",
      "2257 MITE_T_72152|chr4B|172884545|172884708|AT|55|F2976\n",
      "2209 MITE_T_74806|chr6D|218383800|218383961|TA|171|F3083\n",
      "2159 MITE_T_9548|chr7D|550352934|550353282|AAACTA|15|F545\n",
      "2091 MITE_T_88968|chr5A|114487739|114487863|TA|131|F3834\n",
      "2081 MITE_T_117579|chr1D|407351095|407351177|TA|86|F5366\n",
      "2079 MITE_T_67109|chr1B|329075749|329075919|TA|182|F2710\n",
      "2067 MITE_T_83294|chr6B|589204114|589204245|TA|135|F3611\n",
      "2063 MITE_T_90200|chr4A|477881757|477881880|TA|127|F3894\n",
      "2049 MITE_T_89782|chr1D|91443574|91443697|TA|127|F3867\n",
      "2046 MITE_T_88516|chr7A|454048958|454049083|AT|132|F3811\n",
      "2045 MITE_T_9751|chr3D|162181248|162181595|TTA|15|F550\n",
      "2022 MITE_T_117610|chr7A|172464735|172464817|TA|86|F5367\n",
      "2021 MITE_T_10423|chr7A|583289206|583289552|TAA|18|F555\n",
      "2011 MITE_T_91063|chr1B|97377115|97377238|TA|127|F3948\n",
      "1963 MITE_T_69595|chr4B|551079913|551080080|TA|179|F2836\n",
      "1957 MITE_T_76155|chr3B|716631004|716631164|TA|166|F3183\n",
      "1952 MITE_T_91352|chr3D|87925675|87925797|TA|129|F3964\n",
      "1915 MITE_T_74813|chr2A|124381851|124382012|TA|55|F3084\n",
      "1897 MITE_T_70697|chr5D|322384932|322385097|TG|173|F2897\n",
      "1890 MITE_T_87107|chr7B|631752832|631752959|TTTC|129|F3752\n",
      "1870 MITE_T_76212|chr5A|452111601|452111761|TA|172|F3186\n",
      "1868 MITE_T_68427|chr6A|222571242|222571410|AT|177|F2764\n",
      "1862 MITE_T_88037|chr7B|140609405|140609530|AT|129|F3786\n",
      "1860 MITE_T_69040|chr4D|182066918|182067085|AA|174|F2798\n",
      "1846 MITE_T_75137|chr2D|490680145|490680306|AT|172|F3101\n",
      "1811 MITE_T_74852|chr3B|752110244|752110405|TA|173|F3085\n",
      "1786 MITE_T_86573|chr1B|606845299|606845426|TT|127|F3725\n",
      "1773 MITE_T_97305|chr7B|194283260|194283371|AT|29|F4298\n",
      "1751 MITE_T_11589|chr4B|513872438|513872777|AT|46|F594\n",
      "1749 MITE_T_110031|chr2D|206682968|206683058|AT|94|F5010\n",
      "1741 MITE_T_73480|chr2B|52872764|52872927|ATAT|176|F3005\n",
      "1735 MITE_T_75256|chr3D|555180571|555180732|TA|172|F3110\n",
      "1733 MITE_T_90412|chr3D|612269564|612269687|TA|129|F3899\n",
      "1727 MITE_T_70885|chr4A|521301879|521302044|TC|173|F2903\n",
      "1719 MITE_T_75190|chr2D|247099513|247099674|TA|173|F3106\n",
      "1717 MITE_T_85948|chr3A|159530327|159530456|AA|131|F3693\n",
      "1682 MITE_T_76082|chr6D|302057023|302057183|TA|173|F3179\n",
      "1678 MITE_T_73812|chr1D|104950255|104950417|TA|175|F3035\n",
      "1629 MITE_T_89283|chr6B|139746909|139747033|TA|129|F3835\n",
      "1624 MITE_T_75448|chr3A|671642744|671642905|TA|172|F3121\n",
      "1615 MITE_T_76394|chr2A|761249036|761249196|TA|173|F3206\n",
      "1595 MITE_T_89727|chr4D|139419347|139419470|TA|127|F3864\n",
      "1594 MITE_T_73644|chr7B|702173127|702173290|AT|174|F3016\n",
      "1550 MITE_T_111107|chr5A|331394412|331394501|AT|94|F5061\n",
      "1540 MITE_T_113631|chr3D|554346115|554346202|AA|87|F5162\n",
      "1538 MITE_T_90114|chr6D|272859229|272859352|TA|129|F3890\n",
      "1530 MITE_T_83113|chr6B|482153897|482154028|TG|131|F3604\n",
      "1468 MITE_T_87899|chr4D|94261238|94261363|AT|131|F3781\n",
      "1454 MITE_T_64232|chr7D|50710236|50710413|TA|189|F2575\n",
      "1450 MITE_T_49891|chr5B|94517806|94518045|AT|41|F1957\n",
      "1447 MITE_T_91099|chr2B|776888741|776888864|TATA|131|F3950\n",
      "1432 MITE_T_85318|chr6B|143826224|143826353|AC|129|F3661\n",
      "1418 MITE_T_24214|chr4B|466828020|466828318|AGAA|34|F980\n",
      "1415 MITE_T_75048|chr4A|534233761|534233922|TA|172|F3095\n",
      "1385 MITE_T_110904|chr2B|370998573|370998662|TA|95|F5052\n",
      "1380 MITE_T_84171|chr7B|720120585|720120716|AT|137|F3615\n",
      "1367 MITE_T_85687|chr2B|738826100|738826229|TA|136|F3690\n",
      "1366 MITE_T_66395|chr2D|428669394|428669564|TA|178|F2699\n"
     ]
    }
   ],
   "source": [
    "count = 1\n",
    "for k in s.head(100).index.tolist():\n",
    "#    df_blat[df_blat.qName == k].sort_values(by=['tName','tStart']).to_csv(str(count) + '.csv')\n",
    "    count += 1\n",
    "    print(s[k],k)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total genes 110789\n",
      "Genes with MITEs 7767\n",
      "That\"s 7.010623798391537\n"
     ]
    }
   ],
   "source": [
    "#count genes with MITEs in full genome\n",
    "import pandas as pd\n",
    "\n",
    "genome_file = 'extras/data/iwgsc.gff3'\n",
    "df_genome = pd.read_csv(genome_file, index_col=False, sep='\\t', header=None)\n",
    "df_genome.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "\n",
    "\n",
    "df_blat = pd.read_csv('extras/data/iwgsc_cn4/blast_95.csv', index_col=False, sep='\\t')\n",
    "df_blat.columns = ['qseqid','sseqid','pident','length','mismatch','gapopen','qstart','qend','sstart','send','evalue','bitscore']\n",
    "\n",
    "df_genes = pd.read_csv('../data/iwgsc_refseqv1.0_HighConf_2017Mar13.gff3', index_col=False, sep='\\t')\n",
    "df_genes.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes = df_genes[df_genes.feature == 'gene'].reset_index(drop=True)\n",
    "\n",
    "blat_genome = {}\n",
    "for k,chromosome in df_genome.iterrows():\n",
    "    blat_genome[chromosome.seqname] = df_blat[(df_blat.sseqid == chromosome.seqname)]\n",
    "blat_genome[chromosome.seqname] = df_blat[(df_blat.sseqid == chromosome.seqname)]\n",
    "\n",
    "gene_with_mites_count = 0\n",
    "gene_count = len(df_genes.index)\n",
    "gene_names = []\n",
    "for k2,gene in df_genes.iterrows():\n",
    "    curr_df = blat_genome[gene.seqname]\n",
    "    res_1 = curr_df[(curr_df.sstart >= gene.start) & (curr_df.sstart <= gene.end)]\n",
    "    res_2 = curr_df[(curr_df.send >= gene.start) & (curr_df.send <= gene.end)]\n",
    "    if len(res_1.index) > 0 or len(res_2.index) > 0:\n",
    "        gene_names.append(gene.attribute)\n",
    "        gene_with_mites_count += 1\n",
    "\n",
    "gene_names = set(gene_names)\n",
    "thefile = open('extras/genes_with_mites.txt', 'w')\n",
    "for item in gene_names:\n",
    "    thefile.write(\"%s\\n\" % item)\n",
    "\n",
    "print('Total genes', gene_count)\n",
    "print('Genes with MITEs', gene_with_mites_count)\n",
    "print('That\"s',  gene_with_mites_count * 100 / gene_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MITE_T_100|chr4B|149720156|149720951|AT|279|F15</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>98.36</td>\n",
       "      <td>795</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>795</td>\n",
       "      <td>562130598</td>\n",
       "      <td>562129804</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1375.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MITE_T_316|chr1A|335048447|335048919|CT|15|F33</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>100.00</td>\n",
       "      <td>472</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>472</td>\n",
       "      <td>335048448</td>\n",
       "      <td>335048919</td>\n",
       "      <td>0.0</td>\n",
       "      <td>852.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MITE_T_332|chr1A|557861785|557862256|TAA|21|F35</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>100.00</td>\n",
       "      <td>471</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>471</td>\n",
       "      <td>557861786</td>\n",
       "      <td>557862256</td>\n",
       "      <td>0.0</td>\n",
       "      <td>850.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MITE_T_417|chr3A|333269323|333269791|TAGCCCAAG...</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>96.58</td>\n",
       "      <td>468</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>468</td>\n",
       "      <td>259994055</td>\n",
       "      <td>259993598</td>\n",
       "      <td>0.0</td>\n",
       "      <td>773.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MITE_T_547|chr7A|173865867|173866329|TATATATAT...</td>\n",
       "      <td>chr1A</td>\n",
       "      <td>95.89</td>\n",
       "      <td>438</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "      <td>462</td>\n",
       "      <td>100100967</td>\n",
       "      <td>100101404</td>\n",
       "      <td>0.0</td>\n",
       "      <td>706.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              qseqid sseqid  pident  length  \\\n",
       "0    MITE_T_100|chr4B|149720156|149720951|AT|279|F15  chr1A   98.36     795   \n",
       "1     MITE_T_316|chr1A|335048447|335048919|CT|15|F33  chr1A  100.00     472   \n",
       "2    MITE_T_332|chr1A|557861785|557862256|TAA|21|F35  chr1A  100.00     471   \n",
       "3  MITE_T_417|chr3A|333269323|333269791|TAGCCCAAG...  chr1A   96.58     468   \n",
       "4  MITE_T_547|chr7A|173865867|173866329|TATATATAT...  chr1A   95.89     438   \n",
       "\n",
       "   mismatch  gapopen  qstart  qend     sstart       send  evalue  bitscore  \n",
       "0        13        0       1   795  562130598  562129804     0.0    1375.0  \n",
       "1         0        0       1   472  335048448  335048919     0.0     852.0  \n",
       "2         0        0       1   471  557861786  557862256     0.0     850.0  \n",
       "3         6        2       1   468  259994055  259993598     0.0     773.0  \n",
       "4        17        1      26   462  100100967  100101404     0.0     706.0  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_blat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count genes with MITEs per chromosome\n",
    "import pandas as pd\n",
    "\n",
    "df_blat = pd.read_csv('data/iwgsc_cn4/blat_95_95.csv', index_col=False, sep='\\t')\n",
    "\n",
    "df_genes = pd.read_csv('../../data/iwgsc_refseqv1.0_HighConf_2017Mar13.gff3', index_col=False, sep='\\t')\n",
    "df_genes.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes = df_genes[df_genes.feature == 'gene'].reset_index(drop=True)\n",
    "\n",
    "df_genome = pd.read_csv('data/iwgsc.gff3', index_col=False, sep='\\t', header=None)\n",
    "df_genome.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "\n",
    "blat_genome = {}\n",
    "for k1,genome in df_genome.iterrows():\n",
    "    curr_blat = df_blat[(df_blat.tName == genome.seqname)] \n",
    "    curr_genes = df_genes[(df_genes.seqname == genome.seqname)]\n",
    "    gene_with_mites_count = 0\n",
    "    gene_count = len(curr_genes.index)\n",
    "    print(gene_count)\n",
    "    for k2,gene in curr_genes.iterrows():\n",
    "        res_1 = curr_blat[(curr_blat.tStart >= gene.start) & (curr_blat.tStart <= gene.end)]\n",
    "        res_2 = curr_blat[(curr_blat.tEnd >= gene.start) & (curr_blat.tEnd <= gene.end)]\n",
    "        if len(res_1.index) > 0 or len(res_2.index) > 0:\n",
    "            gene_with_mites_count += 1\n",
    "    print(genome.seqname,gene_with_mites_count)\n",
    "    print(genome.seqname,gene_with_mites_count * 100 / gene_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count genes with MITEs per genome\n",
    "import pandas as pd\n",
    "\n",
    "df_blat = pd.read_csv('data/iwgsc_cn4/blat_95_95.csv', index_col=False, sep='\\t')\n",
    "\n",
    "df_genes = pd.read_csv('../../data/iwgsc_refseqv1.0_HighConf_2017Mar13.gff3', index_col=False, sep='\\t')\n",
    "df_genes.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes = df_genes[df_genes.feature == 'gene'].reset_index(drop=True)\n",
    "\n",
    "df_genome = pd.read_csv('data/iwgsc.gff3', index_col=False, sep='\\t', header=None)\n",
    "df_genome.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "print('starting...')\n",
    "blat_genome = {}\n",
    "for k,chromosome in df_genome.iterrows():\n",
    "    blat_genome[chromosome.seqname] = df_blat[(df_blat.tName == chromosome.seqname)]\n",
    "print('proccessing genomes...')\n",
    "for genome in ['A','B','D']:\n",
    "    curr_genes = df_genes[(df_genes.seqname.str.contains(genome))]\n",
    "    gene_with_mites_count = 0\n",
    "    gene_count = len(curr_genes.index)\n",
    "    print(gene_count)\n",
    "    for k2,gene in curr_genes.iterrows():\n",
    "        curr_df = blat_genome[gene.seqname]\n",
    "        res_1 = curr_df[(curr_df.tStart >= gene.start) & (curr_df.tStart <= gene.end)]\n",
    "        res_2 = curr_df[(curr_df.tEnd >= gene.start) &  (curr_df.tEnd <= gene.end)]\n",
    "        if len(res_1.index) > 0 or len(res_2.index) > 0:\n",
    "            if k2 % 1000 == 0:\n",
    "                print(k2,gene_with_mites_count)\n",
    "            gene_with_mites_count += 1\n",
    "    print(genome,gene_with_mites_count)\n",
    "    print(genome,gene_with_mites_count * 100 / gene_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a\n",
      "chr1A 0.15468318796728756\n",
      "chr1B 0.19118104296796354\n",
      "chr1D 0.1730583280576583\n",
      "chr2A 0.1511612680913292\n",
      "chr2B 0.20055644713068022\n",
      "chr2D 0.17805129318735302\n",
      "chr3A 0.151991964867668\n",
      "chr3B 0.192896315158974\n",
      "chr3D 0.16570757613604584\n",
      "chr4A 0.13973442771263417\n",
      "chr4B 0.17963250684495652\n",
      "chr4D 0.15141361176817816\n",
      "chr5A 0.14981492489501685\n",
      "chr5B 0.20531409926568903\n",
      "chr5D 0.16475955422869876\n",
      "chr6A 0.13810219097142978\n",
      "chr6B 0.18103493187862013\n",
      "chr6D 0.15735714922880212\n",
      "chr7A 0.148713143239906\n",
      "chr7B 0.203842052597599\n",
      "chr7D 0.16337760811139052\n",
      "A 0.14774301539218165\n",
      "B 0.19349391369206895\n",
      "D 0.1648178743883038\n",
      "Total 0.16868493449085142\n"
     ]
    }
   ],
   "source": [
    "#percentage of genome covered by MITEs\n",
    "import pandas as pd\n",
    "\n",
    "def merge_overlap(intervals):\n",
    "    sorted_by_lower_bound = sorted(intervals, key=lambda tup: tup[0])\n",
    "    merged = []\n",
    "    for higher in sorted_by_lower_bound:\n",
    "        if not merged:\n",
    "            merged.append(higher)\n",
    "        else:\n",
    "            lower = merged[-1]\n",
    "            # test for intersection between lower and higher:\n",
    "            # we know via sorting that lower[0] <= higher[0]\n",
    "            if higher[0] <= lower[1]:\n",
    "                upper_bound = max(lower[1], higher[1])\n",
    "                merged[-1] = (lower[0], upper_bound)  # replace by merged interval\n",
    "            else:\n",
    "                merged.append(higher)\n",
    "    return merged\n",
    "\n",
    "df_blat = pd.read_csv('extras/data/iwgsc_cn4/blast_95.csv', index_col=False, sep='\\t')\n",
    "\n",
    "df_genome = pd.read_csv('extras/data/iwgsc.gff3', index_col=False, sep='\\t', header=None)\n",
    "df_genome.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "intervals = {}\n",
    "for k,v in df_blat.iterrows():\n",
    "    genome = v.sseqid\n",
    "    if not genome in intervals:\n",
    "        intervals[genome] = []\n",
    "    intervals[genome].append((v.sstart,v.send))\n",
    "    \n",
    "print(\"a\")\n",
    "percs_genome = {}\n",
    "total_sum = 0\n",
    "for k,v in df_genome.iterrows():\n",
    "    genome = v.seqname[4:6]\n",
    "    if not genome in percs_genome:\n",
    "        percs_genome[genome] = 0\n",
    "    current_interval = merge_overlap(intervals[v.seqname])\n",
    "    interval_sum = 0\n",
    "    for i in current_interval:\n",
    "        interval_sum += max(i) - min(i)\n",
    "    perc = interval_sum * 100 / v.end\n",
    "    print(v.seqname, perc)\n",
    "    #calculate avg for full genomes\n",
    "    percs_genome[genome] += perc\n",
    "    total_sum += perc\n",
    "    \n",
    "for genome in ['A','B','D']:\n",
    "    print(genome, percs_genome[genome] / 7)\n",
    "print('Total', total_sum / 21)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count genes per genome\n",
    "import pandas as pd\n",
    "df_genes = pd.read_csv('../../data/iwgsc_refseqv1.0_HighConf_2017Mar13.gff3', index_col=False, sep='\\t')\n",
    "df_genes.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes = df_genes[df_genes.feature == 'gene']\n",
    "total_genes = len(df_genes.index)\n",
    "print('total genes', total_genes)\n",
    "s = df_genes.seqname.value_counts()\n",
    "genome = {}\n",
    "for k in s.index.tolist():\n",
    "    gen = k[4:6]\n",
    "    if not gen in genome:\n",
    "        genome[gen] = s[k]\n",
    "    else:\n",
    "        genome[gen] += s[k]\n",
    "print(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate MITE and gene density\n",
    "import pandas as pd\n",
    "import sys\n",
    "df_blast = pd.read_csv('data/iwgsc_cn4/blat_95_95.csv', index_col=False, sep='\\t')\n",
    "df_genome = pd.read_csv('data/iwgsc.gff3', index_col=False, sep='\\t', header=None)\n",
    "df_genome.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "total_bp = 0\n",
    "for k,v in df_genome.iterrows():\n",
    "    total_bp += v.end\n",
    "\n",
    "total_mites = len(df_blast.index)\n",
    "total_mb = total_bp / 1000 / 1000\n",
    "total_mb_total = total_mb\n",
    "total_gb = total_mb / 1000 \n",
    "print ('Total:','&',int(total_mb),'&',total_mites,'&',int(total_mites / total_mb))\n",
    "\n",
    "genome_sizes = {}\n",
    "for k,v in df_genome.iterrows():\n",
    "    df_blast_seq = df_blast[df_blast.tName == v.seqname]\n",
    "    total_mites = len(df_blast_seq.index)\n",
    "    total_mb = v.end / 1000 / 1000\n",
    "    seq = v.seqname.replace('chr','')\n",
    "    if not seq[1:2] in genome_sizes:\n",
    "        genome_sizes[seq[1:2]] = total_mb\n",
    "    else:\n",
    "        genome_sizes[seq[1:2]] += total_mb\n",
    "    print (seq,'&',int(total_mb),'&',total_mites,'&',int(total_mites / total_mb))\n",
    "\n",
    "for genome in ['A','B','D']:\n",
    "    df_blast_seq = df_blast[df_blast.tName.str.contains(genome)]\n",
    "    total_mites = len(df_blast_seq.index)\n",
    "    print (genome,'&', int(genome_sizes[genome]),'&',total_mites,'&',int(total_mites / genome_sizes[genome]))\n",
    "print('*' * 10)\n",
    "    \n",
    "    \n",
    "df_genes = pd.read_csv('../../data/iwgsc_refseqv1.0_HighConf_2017Mar13.gff3', index_col=False, sep='\\t')\n",
    "df_genes.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes = df_genes[df_genes.feature == 'gene']\n",
    "total_genes = len(df_genes.index)\n",
    "print ('Total Genes:',total_genes)\n",
    "print ('Total Genes per MB:',total_genes / total_mb_total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
