{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count genes with MITEs in full genome\n",
    "import pandas as pd\n",
    "\n",
    "df_blat = pd.read_csv('data/iwgsc_cn4/blat_95_95.csv', index_col=False, sep='\\t')\n",
    "\n",
    "df_genes = pd.read_csv('../../data/iwgsc_refseqv1.0_HighConf_2017Mar13.gff3', index_col=False, sep='\\t')\n",
    "df_genes.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes = df_genes[df_genes.feature == 'gene'].reset_index(drop=True)\n",
    "\n",
    "blat_genome = {}\n",
    "for k,chromosome in df_genome.iterrows():\n",
    "    blat_genome[chromosome.seqname] = df_blat[(df_blat.tName == chromosome.seqname)]\n",
    "blat_genome['chrUn'] = df_blat[(df_blat.tName == 'chrUn')]\n",
    "gene_with_mites_count = 0\n",
    "gene_count = len(df_genes.index)\n",
    "for k2,gene in df_genes.iterrows():\n",
    "    curr_df = blat_genome[gene.seqname]\n",
    "    res_1 = curr_df[(curr_df.tStart >= gene.start) & (curr_df.tStart <= gene.end)]\n",
    "    res_2 = curr_df[(curr_df.tEnd >= gene.start) & (curr_df.tEnd <= gene.end)]\n",
    "    if len(res_1.index) > 0 or len(res_2.index) > 0:\n",
    "        gene_with_mites_count += 1\n",
    "print(gene_with_mites_count)\n",
    "print(gene_with_mites_count * 100 / gene_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count genes with MITEs per chromosome\n",
    "import pandas as pd\n",
    "\n",
    "df_blat = pd.read_csv('data/iwgsc_cn4/blat_95_95.csv', index_col=False, sep='\\t')\n",
    "\n",
    "df_genes = pd.read_csv('../../data/iwgsc_refseqv1.0_HighConf_2017Mar13.gff3', index_col=False, sep='\\t')\n",
    "df_genes.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes = df_genes[df_genes.feature == 'gene'].reset_index(drop=True)\n",
    "\n",
    "df_genome = pd.read_csv('data/iwgsc.gff3', index_col=False, sep='\\t', header=None)\n",
    "df_genome.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "\n",
    "blat_genome = {}\n",
    "for k1,genome in df_genome.iterrows():\n",
    "    curr_blat = df_blat[(df_blat.tName == genome.seqname)] \n",
    "    curr_genes = df_genes[(df_genes.seqname == genome.seqname)]\n",
    "    gene_with_mites_count = 0\n",
    "    gene_count = len(curr_genes.index)\n",
    "    print(gene_count)\n",
    "    for k2,gene in curr_genes.iterrows():\n",
    "        res_1 = curr_blat[(curr_blat.tStart >= gene.start) & (curr_blat.tStart <= gene.end)]\n",
    "        res_2 = curr_blat[(curr_blat.tEnd >= gene.start) & (curr_blat.tEnd <= gene.end)]\n",
    "        if len(res_1.index) > 0 or len(res_2.index) > 0:\n",
    "            gene_with_mites_count += 1\n",
    "    print(genome.seqname,gene_with_mites_count)\n",
    "    print(genome.seqname,gene_with_mites_count * 100 / gene_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count genes with MITEs per genome\n",
    "import pandas as pd\n",
    "\n",
    "df_blat = pd.read_csv('data/iwgsc_cn4/blat_95_95.csv', index_col=False, sep='\\t')\n",
    "\n",
    "df_genes = pd.read_csv('../../data/iwgsc_refseqv1.0_HighConf_2017Mar13.gff3', index_col=False, sep='\\t')\n",
    "df_genes.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes = df_genes[df_genes.feature == 'gene'].reset_index(drop=True)\n",
    "\n",
    "df_genome = pd.read_csv('data/iwgsc.gff3', index_col=False, sep='\\t', header=None)\n",
    "df_genome.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "print('starting...')\n",
    "blat_genome = {}\n",
    "for k,chromosome in df_genome.iterrows():\n",
    "    blat_genome[chromosome.seqname] = df_blat[(df_blat.tName == chromosome.seqname)]\n",
    "print('proccessing genomes...')\n",
    "for genome in ['A','B','D']:\n",
    "    curr_genes = df_genes[(df_genes.seqname.str.contains(genome))]\n",
    "    gene_with_mites_count = 0\n",
    "    gene_count = len(curr_genes.index)\n",
    "    print(gene_count)\n",
    "    for k2,gene in curr_genes.iterrows():\n",
    "        curr_df = blat_genome[gene.seqname]\n",
    "        res_1 = curr_df[(curr_df.tStart >= gene.start) & (curr_df.tStart <= gene.end)]\n",
    "        res_2 = curr_df[(curr_df.tEnd >= gene.start) &  (curr_df.tEnd <= gene.end)]\n",
    "        if len(res_1.index) > 0 or len(res_2.index) > 0:\n",
    "            if k2 % 1000 == 0:\n",
    "                print(k2,gene_with_mites_count)\n",
    "            gene_with_mites_count += 1\n",
    "    print(genome,gene_with_mites_count)\n",
    "    print(genome,gene_with_mites_count * 100 / gene_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#percentage of genome covered by MITEs\n",
    "import pandas as pd\n",
    "\n",
    "def merge_overlap(intervals):\n",
    "    sorted_by_lower_bound = sorted(intervals, key=lambda tup: tup[0])\n",
    "    merged = []\n",
    "    for higher in sorted_by_lower_bound:\n",
    "        if not merged:\n",
    "            merged.append(higher)\n",
    "        else:\n",
    "            lower = merged[-1]\n",
    "            # test for intersection between lower and higher:\n",
    "            # we know via sorting that lower[0] <= higher[0]\n",
    "            if higher[0] <= lower[1]:\n",
    "                upper_bound = max(lower[1], higher[1])\n",
    "                merged[-1] = (lower[0], upper_bound)  # replace by merged interval\n",
    "            else:\n",
    "                merged.append(higher)\n",
    "    return merged\n",
    "\n",
    "df_blat = pd.read_csv('data/iwgsc_cn4/blat_95_95.csv', index_col=False, sep='\\t')\n",
    "\n",
    "df_genome = pd.read_csv('data/iwgsc.gff3', index_col=False, sep='\\t', header=None)\n",
    "df_genome.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "intervals = {}\n",
    "for k,v in df_blat.iterrows():\n",
    "    genome = v.tName\n",
    "    if not genome in intervals:\n",
    "        intervals[genome] = []\n",
    "    intervals[genome].append((v.tStart,v.tEnd))\n",
    "    \n",
    "print(\"a\")\n",
    "percs_genome = {}\n",
    "total_sum = 0\n",
    "for k,v in df_genome.iterrows():\n",
    "    genome = v.seqname[4:6]\n",
    "    if not genome in percs_genome:\n",
    "        percs_genome[genome] = 0\n",
    "    current_interval = merge_overlap(intervals[v.seqname])\n",
    "    interval_sum = 0\n",
    "    for i in current_interval:\n",
    "        interval_sum += max(i) - min(i)\n",
    "    perc = interval_sum * 100 / v.end\n",
    "    print(v.seqname, perc)\n",
    "    #calculate avg for full genomes\n",
    "    percs_genome[genome] += perc\n",
    "    total_sum += perc\n",
    "    \n",
    "for genome in ['A','B','D']:\n",
    "    print(genome, percs_genome[genome] / 7)\n",
    "print('Total', total_sum / 21)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count genes per genome\n",
    "import pandas as pd\n",
    "df_genes = pd.read_csv('../../data/iwgsc_refseqv1.0_HighConf_2017Mar13.gff3', index_col=False, sep='\\t')\n",
    "df_genes.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes = df_genes[df_genes.feature == 'gene']\n",
    "total_genes = len(df_genes.index)\n",
    "print('total genes', total_genes)\n",
    "s = df_genes.seqname.value_counts()\n",
    "genome = {}\n",
    "for k in s.index.tolist():\n",
    "    gen = k[4:6]\n",
    "    if not gen in genome:\n",
    "        genome[gen] = s[k]\n",
    "    else:\n",
    "        genome[gen] += s[k]\n",
    "print(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate MITE and gene density\n",
    "import pandas as pd\n",
    "import sys\n",
    "df_blast = pd.read_csv('data/iwgsc_cn4/blat_95_95.csv', index_col=False, sep='\\t')\n",
    "df_genome = pd.read_csv('data/iwgsc.gff3', index_col=False, sep='\\t', header=None)\n",
    "df_genome.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "total_bp = 0\n",
    "for k,v in df_genome.iterrows():\n",
    "    total_bp += v.end\n",
    "\n",
    "total_mites = len(df_blast.index)\n",
    "total_mb = total_bp / 1000 / 1000\n",
    "total_mb_total = total_mb\n",
    "total_gb = total_mb / 1000 \n",
    "print ('Total:','&',int(total_mb),'&',total_mites,'&',int(total_mites / total_mb))\n",
    "\n",
    "genome_sizes = {}\n",
    "for k,v in df_genome.iterrows():\n",
    "    df_blast_seq = df_blast[df_blast.tName == v.seqname]\n",
    "    total_mites = len(df_blast_seq.index)\n",
    "    total_mb = v.end / 1000 / 1000\n",
    "    seq = v.seqname.replace('chr','')\n",
    "    if not seq[1:2] in genome_sizes:\n",
    "        genome_sizes[seq[1:2]] = total_mb\n",
    "    else:\n",
    "        genome_sizes[seq[1:2]] += total_mb\n",
    "    print (seq,'&',int(total_mb),'&',total_mites,'&',int(total_mites / total_mb))\n",
    "\n",
    "for genome in ['A','B','D']:\n",
    "    df_blast_seq = df_blast[df_blast.tName.str.contains(genome)]\n",
    "    total_mites = len(df_blast_seq.index)\n",
    "    print (genome,'&', int(genome_sizes[genome]),'&',total_mites,'&',int(total_mites / genome_sizes[genome]))\n",
    "print('*' * 10)\n",
    "    \n",
    "    \n",
    "df_genes = pd.read_csv('../../data/iwgsc_refseqv1.0_HighConf_2017Mar13.gff3', index_col=False, sep='\\t')\n",
    "df_genes.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df_genes = df_genes[df_genes.feature == 'gene']\n",
    "total_genes = len(df_genes.index)\n",
    "print ('Total Genes:',total_genes)\n",
    "print ('Total Genes per MB:',total_genes / total_mb_total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
